File: 7-reverse-integer/reverse-integer.go
func reverse(x int) int {
    rev := 0
    mul := 1
    if x < 0 {
        mul = -1
        x = x * -1
    }

    for x > 0 {
        digit := x % 10
        rev = rev * 10 + digit
        x = x / 10
    }

    if (rev >> 31) == 0 {
        return rev * mul
    }

    return 0
}

File: 7-reverse-integer/reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        sign = 1 
        if x < 0:
            x = -x
            sign = -1

        rev = 0
        MAX_INT = (1 << 31) - 1

        while x:
            digit = int(x % 10)
            
            if rev > MAX_INT // 10 or (rev == MAX_INT // 10 and digit == 7):
                return 0
            
            rev = rev * 10 + digit
            x //= 10
        
        rev *= sign
        return rev if -(1 << 31) <= rev <= MAX_INT else 0
